velox/type/DecimalUtil.h

```diff
@@ -203,29 +204,55 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
-  /// Rescales a double value to decimal value of given precision and scale. The
-  /// output is rescaled value of int128_t or int64_t type. Returns error status
-  /// if fails.
-  template <typename TOutput>
-  inline static Status
-  rescaleDouble(double value, int precision, int scale, TOutput& output) {
+  /// Rescales a floating point value to decimal value of given precision and
+  /// scale. The output is rescaled value of int128_t or int64_t type. Returns
+  /// error status if fails.
+  template <typename TIntput, typename TOutput>
+  inline static Status rescaleFloatingPoint(
+      TIntput value,
+      int precision,
+      int scale,
+      TOutput& output) {
     if (!std::isfinite(value)) {
       return Status::UserError("The input value should be finite.");
     }
 
+    uint8_t digits;
+    if constexpr (std::is_same_v<TIntput, float>) {
+      // A float provides between 6 and 7 decimal digits, so at least 6 digits
+      // are precise.
+      digits = 6;
+    } else {
+      // A double provides from 15 to 17 decimal digits, so at least 15 digits
+      // are precise.
+      digits = 15;
+      if (value <= std::numeric_limits<int128_t>::min() ||
+          value >= std::numeric_limits<int128_t>::max()) {
+        return Status::UserError("Result overflows.");
+      }
+    }
+
+    // Calculate the precise fractional digits.
+    const auto integralValue =
+        static_cast<uint128_t>(value > 0 ? value : -value);
+    const auto integralDigits =
```

Comment:
For previous Cast(double as decimal), this implement introduce extra integralDigits computation

Reply:
This part is needed for both double and float as previous implementation brings some precision issue. I will explain in detail in the PR description.

Reply:
Updated. Thanks.